packages/dds/tree/src/test/feature-libraries/default-schema/schemaChecker.spec.ts

```diff
@@ -0,0 +1,722 @@
+					numberNode.type,
+				]);
+				const nodeSchema_object: TreeNodeStoredSchema = new ObjectNodeStoredSchema(
+					new Map([[brand("requiredProp"), fieldSchema_requiredNumber]]),
+				);
+
+				// Create an object node with no fields at all; particularly it doesn't have the required 'requiredProp' field
+				const objectNode = getObjectNode("myObjectNode", new Map([]));
+
+				const schemaCollection: StoredSchemaCollection = {
+					nodeSchema: new Map([
+						[numberNode.type, new LeafNodeStoredSchema(ValueSchema.Number)],
+						[objectNode.type, nodeSchema_object],
+					]),
+				};
+				const schemaPolicy = {
+					fieldKinds: new Map([[fieldSchema_requiredNumber.kind, FieldKinds.required]]),
+				};
+
+				assert.equal(
+					isNodeInSchema(objectNode, {
+						schema: schemaCollection,
+						policy: schemaPolicy,
+					}),
+					SchemaValidationErrors.Field_IncorrectMultiplicity,
+				);
+			});
+		});
+	});
+
+	describe("isFieldInSchema", () => {
+		it(`not in schema if field kind not supported by schema policy`, () => {
+			const numberNode = getValueNode("myNumberNode", 1);
+			const fieldSchema = getFieldSchema(FieldKinds.required, [numberNode.type]);
+			const schemaCollection = {
+				nodeSchema: new Map([
+					[numberNode.type, new LeafNodeStoredSchema(ValueSchema.Number)],
+				]),
+			};
+
+			// FieldKinds.required is used above but missing in the schema policy
+			assert.equal(
+				isFieldInSchema([numberNode], fieldSchema, {
+					schema: schemaCollection,
+					policy: emptySchemaPolicy,
+				}),
+				SchemaValidationErrors.Field_KindNotInSchemaPolicy,
+			);
+		});
+
+		it(`not in schema if type of a child node is not supported by field`, () => {
+			const numberNode = getValueNode("myNumberNode", 1);
+			const stringNode = getValueNode("myStringNode", "myStringValue");
+			const fieldSchema = getFieldSchema(FieldKinds.sequence, [numberNode.type]);
+			const schemaCollection = {
+				nodeSchema: new Map([
+					[numberNode.type, new LeafNodeStoredSchema(ValueSchema.Number)],
+				]),
+			};
+			const schemaPolicy = {
+				fieldKinds: new Map([[fieldSchema.kind, FieldKinds.sequence]]),
+			};
+
+			// Confirm that the field supports number nodes
+			assert.equal(
+				isFieldInSchema([numberNode], fieldSchema, {
+					schema: schemaCollection,
+					policy: schemaPolicy,
+				}),
+				SchemaValidationErrors.NoError,
+			);
+
+			// Field does not support string nodes
+			assert.equal(
+				isFieldInSchema([stringNode], fieldSchema, {
+					schema: schemaCollection,
+					policy: schemaPolicy,
+				}),
+				SchemaValidationErrors.Field_NodeTypeNotAllowed,
+			);
+
+			// Still fails even if there are other valid nodes for the field
+			assert.equal(
+				isFieldInSchema([numberNode, stringNode, numberNode], fieldSchema, {
+					schema: schemaCollection,
+					policy: schemaPolicy,
+				}),
+				SchemaValidationErrors.Field_NodeTypeNotAllowed,
+			);
+		});
+
+		const isFieldInSchema_multiplicityTestCases: [
```

Comment:
you already test compliesWithMultiplicity above, so here you only really need to test 2 cases: correct and incorrect multiplicity.

Reply:
I agree it feels redundant, but it was intentional thinking that if at some point `isFieldInSchema()` stops leveraging `compliesWithMultiplicity()` for some reason, we'd still want to ensure that it doesn't start missing one case by mistake. The test is not particularly expensive and the cases are few, so I thought the tradeoff was acceptable. Don't have particularly strong feelings about it thought, if you think it's preferable to just pick one success and one failure case from among the possible inputs, I'm fine doing that.

Reply:
I like to think of there being two approaches here: we can test that each chunk of code is correct assuming its dependencies are correct (fine grained glass box testing) or we can test that each function implements its documented API in some selected test cases.

For code that trivially wraps another function in the same file, taking the second approach, and testing in the same level of fidelity as you test the wrapped function tends to make the size of the test suite scale worse that the size of the codebase resulting it testing harming maintainability instead of improving it.

This entire block of tests is covering that a single chunk of logic, a call to an already tested function, is implemented correctly.

Personally the main value I see from splitting compliesWithMultiplicity into its own function is that we get separation of concerns which is great for testing. If we want to duplicate its test suite which is larger than the implementation code wise, that kinda defeats the point.

You are correct though, testing that the higher level function is fully robust in a self-contained way has value, it just also has cost and this cost seems a little high since I think the value is pretty low. Regardless both the cost and value here are subjective, and its a close call, so feel free to keep this as if you want to.

Reply:
I simplified them into just a test case for not-in-schema due to incorrect multiplicity, and added the explicit test for in-schema for isFieldInSchema.